sdk/iot/pnp/src/az_iot_pnp_client_telemetry.c

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "az_iot_pnp_client.h"
+#include <az_precondition.h>
+#include <az_precondition_internal.h>
+#include <az_result.h>
+#include <az_span.h>
+
+#include <_az_cfg.h>
+
+static const uint8_t pnp_telemetry_param_separator = '&';
+static const uint8_t pnp_telemetry_param_equals = '=';
+
+static const az_span pnp_telemetry_component_name_param = AZ_SPAN_LITERAL_FROM_STR("%24.ifname");
+static const az_span pnp_telemetry_content_type_param = AZ_SPAN_LITERAL_FROM_STR("%24.ct");
+static const az_span pnp_telemetry_content_encoding_param = AZ_SPAN_LITERAL_FROM_STR("%24.ce");
+
+static az_result az_add_telemetry_property(
+    az_span mqtt_topic,
+    az_span property_name,
+    az_span property_value,
+    bool add_separator,
+    az_span* out_mqtt_topic)
+{
+  if (add_separator)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(mqtt_topic, pnp_telemetry_param_separator, &mqtt_topic));
+  }
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic, property_name, &mqtt_topic));
+  AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_topic, pnp_telemetry_param_equals, &mqtt_topic));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic, property_value, &mqtt_topic));
+
+  *out_mqtt_topic = mqtt_topic;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_telemetry_get_publish_topic(
+    az_iot_pnp_client const* client,
+    az_span component_name,
+    az_span mqtt_topic,
+    void* reserved,
+    az_span* out_mqtt_topic)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(component_name, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic, 0, false);
+  AZ_PRECONDITION_IS_NULL(reserved);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic);
+
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_telemetry_publish_topic_get(
+      &client->_internal.iot_hub_client, NULL, mqtt_topic, &mqtt_topic));
+
+  AZ_RETURN_IF_FAILED(az_add_telemetry_property(
+      mqtt_topic, pnp_telemetry_component_name_param, component_name, false, &mqtt_topic));
```

Comment:
Correct me if I am wrong here, but setting the `separator` boolean here to false assumes that the `telemetry_public_topic_get()` API didn't append any properties correct? Because if it did and this prop was added, wouldn't there would be a missing `&`?

Reply:
Yes.  Because I'm passing `properties` as NULL in my call to the topic_get, it's not going to append anything.  I have to make this assumption to get the topic built up correctly.

If on the off-shot you added additional properties for whatever reason (which I don't think you can given the protocol), at a minimum the UT for PnP telemetry would start failing and at that point we'd make required changes.

Reply:
From the `pnp_init()`: 
https://github.com/Azure/azure-sdk-for-c/blob/59d2c6ca8e7ffb10999485bd4bf40992686b817f/sdk/iot/pnp/src/az_iot_pnp_client.c#L42-L46

I think if the `user_agent` is appended that could screw this up?

Reply:
Great catch!  By the way, @danewalton the IoTHub telemetry tests are also missing a test for `user_agent`.

Reply:
I think IIRC telemetry doesn't have it. Oh wait yea it's only on the connect that the user agent is put there. Let me find the PR in which we removed it.

Reply:
https://github.com/Azure/azure-sdk-for-c/pull/487
So in this case I think you're actually okay. My apologies I forgot this was removed in telemetry

Reply:
No need to apologize almost got me, too.  Closing out conversation based on this...